proposals/p1928.md

```diff
@@ -0,0 +1,163 @@
+# Arrays
+
+<!--
+Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+[Pull request](https://github.com/carbon-language/carbon-lang/pull/1928)
+
+<!-- toc -->
+
+## Table of contents
+
+-   [Problem](#problem)
+-   [Background](#background)
+-   [Proposal](#proposal)
+-   [Details](#details)
+-   [Rationale](#rationale)
+-   [Alternatives considered](#alternatives-considered)
+
+<!-- tocstop -->
+
+## Problem
+
+Arrays are base component of almost of programming languages. In the current
+stage of Carbon development, arrays are missed.
+
+## Background
+
+An array is a set of scalar data placed in contiguous memory locations, all of
+the same type and type parameters. The scalar data that make up an array are the
+array elements which can be referenced by unique indexes.
+
+A nested array is an array of arrays which may be non-continuous in memory.
+
+A multidimensional array is an array with more than two dimensions which is
+continuous in memory.
+
+## Proposal
+
+This proposal introduces array internal representation, its declaration, and
+indexing.
+
+## Details
+
+### Internal representation
+
+For simplifying usage and raising safety of array usage, internal representation
+of arrays contains additional metadata:
+```cpp
+  struct {
```

Comment:
OK, then maybe it could be something like this instead:

```cpp
struct {
  std::vector<int32_t> dimension_sizes;
  std::unique_ptr<T[]> data;
}
```

That seems more self-documenting, at least for a C++ programmer. For example, you don't need to explain that `dimension_sizes` is an array, because the type tells you that.

Reply:
I changed it to:
```cpp
template <typename T, int32_t N>
struct Array {
  std::array<int32_t, N> dimension_sizes;
  std::unique_ptr<T[]> data;
};
```

Number of dimensions for arrays shouldn't be changed.